### PR TITLE
[SPARK-18570][ML][R] RFormula support * and ^ operators

### PR DESCRIPTION
**Pull request by ozancicek**
## What changes were proposed in this pull request?

Added support for `*` and `^` operators, along with expressions within parentheses. New operators just expand to already supported terms, such as;

 - y ~ a * b = y ~ a + b + a : b
 - y ~ (a+b+c)^3 = y ~ a + b + c + a : b + a : c + a :b : c 

## How was this patch tested?

Added new unit tests to RFormulaParserSuite

@mengxr @yanboliang

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala [hunk 1]
```diff
@@ -105,12 +112,14 @@ private[feature] trait RFormulaBase extends HasFeaturesCol with HasLabelCol with
...
+    "stringIndexerOrderType",
     "How to order categories of a string FEATURE column used by StringIndexer. " +
-    "The last category after ordering is dropped when encoding strings. " +
-    s"Supported options: ${StringIndexer.supportedStringOrderType.mkString(", ")}. " +
-    "The default value is 'frequencyDesc'. When the ordering is set to 'alphabetDesc', " +
-    "RFormula drops the same category as R when encoding strings.",
+      "The last category after ordering is dropped when encoding strings. " +
+      s"Supported options: ${StringIndexer.supportedStringOrderType.mkString(", ")}. " +
+      "The default value is 'frequencyDesc'. When the ordering is set to 'alphabetDesc', " +
+      "RFormula drops the same category as R when encoding strings.",
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala]: The changes here and above don't seem directly relevant unless I'm missing something. I wouldn't reformat these.

File: mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala [hunk 2]
```diff
@@ -297,13 +315,15 @@ class RFormula @Since("1.5.0") (@Since("1.5.0") override val uid: String)
   @Since("1.5.0")
   // optimistic schema; does not contain any ML attributes
   override def transformSchema(schema: StructType): StructType = {
-    require(!hasLabelCol(schema) || !$(forceIndexLabel),
+    require(
+      !hasLabelCol(schema) || ! $(forceIndexLabel),
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala]: Likewise around here I wouldn't reformat; it's pretty fine or at least consistent with the rest of the code

**Review by srowen — CHANGES_REQUESTED** [review]
etc etc. I'd revert all the changes that aren't directly relevant here to make it easier to review please.

**Review by felixcheung — COMMENTED** [review]
I'd agree, lots of unrelated formatting changes.

also consider R doc has a number of list of operators hard-coded and likely should be updated
https://github.com/apache/spark/search?l=R&q=including+%27~%27%2C+%27.%27

File: mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala (hunk 1 above)
Comment (ozancicek) [reply mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala]: reverted format changes.

File: mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala (hunk 2 above)
Comment (ozancicek) [reply mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala]: reverted format changes

File: mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala [hunk 3]
```diff
@@ -144,10 +155,50 @@ private[ml] case class ResolvedRFormula(
...
+  /**
+   * Fold by adding deletion terms to the left. Double negation
+   * doesn't cancel deletion in order not to add extra terms, e.g.
+   * a - (b - c) = a - Deletion(b) - Deletion(c) = a
+   */
+  def subtract(other: Term): Term = {
+    other.asTerms.terms.foldLeft(this) {
+      case (left, right) =>
+        right match {
+          case t: Deletion => left add t
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala]: Nit: we generally avoid postfix syntax as it generates a compiler warning in some cases. Just `left.add(t)` etc

File: mllib/src/test/scala/org/apache/spark/ml/feature/RFormulaParserSuite.scala [hunk 4]
```diff
@@ -90,12 +90,48 @@ class RFormulaParserSuite extends SparkFunSuite {
...
+  }
+
+  test("interaction distributive") {
+    checkParse("y ~ (a + b):c", "y", Seq("a:c", "b:c"))
+    checkParse("y ~ c:(a + b)", "y", Seq("c:a", "c:b"))
+  }
+
+  test("factor cross distributive") {
+    checkParse("y ~ (a + b)*c", "y", Seq("a", "b", "c", "a:c", "b:c"))
+    checkParse("y ~ c*(a + b)", "y", Seq("c", "a", "b", "c:a", "c:b"))
```
Comment (srowen) [inline mllib/src/test/scala/org/apache/spark/ml/feature/RFormulaParserSuite.scala]: I don't think this is a problem, just curious: the order of terms in an interaction doesn't matter, but I wonder if we handle the case that say a:c and c:a are generated, and only return a:c? might be worth one test case somewhere.

File: mllib/src/test/scala/org/apache/spark/ml/feature/RFormulaParserSuite.scala [hunk 5]
```diff
@@ -90,12 +90,48 @@ class RFormulaParserSuite extends SparkFunSuite {
...
+  }
+
+  test("parse power") {
+    val schema = (new StructType)
+      .add("a", "int", true)
+      .add("b", "long", false)
+      .add("c", "string", true)
+      .add("d", "string", true)
+    checkParse("a ~ (a + b)^2", "a", Seq("a", "b", "a:b"))
```
Comment (srowen) [inline mllib/src/test/scala/org/apache/spark/ml/feature/RFormulaParserSuite.scala]: Finally just wondering about corner cases... what does `(a+b)^1` or `^0` give?

**Review by srowen — COMMENTED** [review]
Looking pretty reasonable, though I don't know this part well. It's just supporting some additional syntactic sugar

File: mllib/src/test/scala/org/apache/spark/ml/feature/RFormulaParserSuite.scala (hunk 5 above)
Comment (ozancicek) [reply mllib/src/test/scala/org/apache/spark/ml/feature/RFormulaParserSuite.scala]: The regex match of ^ operator (line 287 on RFormulaParser.scala) is like this;

`term ~ "^" ~ "^[1-9]\\d*".r`

It only matches numbers larger than 0. So if there is a term with `^0`, directly an exception is raised with `Could not parse formula` message. `^1` terms returns itself, so `(a+b)^1=a+b`

File: mllib/src/test/scala/org/apache/spark/ml/feature/RFormulaParserSuite.scala (hunk 4 above)
Comment (ozancicek) [reply mllib/src/test/scala/org/apache/spark/ml/feature/RFormulaParserSuite.scala]: Yes that's right, if we end up generating `a:c` and `c:a`, only `a:c` should return. I've added some checks for that on lines 93 and 102 on RFormulaParserSuite.scala. Do you have additional test cases in mind or are those good enough?

File: mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala (hunk 3 above)
Comment (ozancicek) [reply mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala]: ok, I got rid of those syntax

**Review by srowen — APPROVED** [review]

**Comment by felixcheung** [comment]
LGTM

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**